### ISSUE TITLE
Change username method?

### ISSUE THREAD

**Issue by turbopixel** [delight-im/PHP-Auth#172]
I want to allow my user to login by username. Currently the column 'username' is empty.

How can I change the user name so that the value in the session is also updated?

**Comment by ocram** [comment]
Good question!

The fact that your `username` column is empty means that you didn’t collect and store a username during registration, right?

If you want to let users change (or add) their usernames later, the corresponding feature for this is unfortunately not available yet – but it can be easily added on top of this library.

So if you want to implement `changeUsername` (probably as a function in your own application code), you can proceed as follows:

 1. Check if the user is already logged in, i.e. whether `Auth#isLoggedIn()` equals `true`.
 1. Using `Auth#getUserId()`, in the database table `users`, for the row with the relevant user ID, update the column `username` to store your new value.
 1. Update `$_SESSION[\Delight\Auth\UserManager::SESSION_FIELD_USERNAME]` to store the name value as well – but only if the previous step succeeded, of course.

If you want the newly implemented function to behave like something you would call `changeUniqueUsername` instead, you would probably want to copy some code from [here](https://github.com/delight-im/PHP-Auth/blob/54f6c5320a7103559937b09ff436a410c756171a/src/UserManager.php#L141).

Hope that helps!

Let’s also leave this issue open until the feature is part of this library itself.

**Comment by turbopixel** [comment]
Correct, user can register without a username. The username column are always empty.

So I create a custom method like this:

```
protected function fieldUsername(string $name) {

    if($this->controller->auth->isLoggedIn() !== true){
        return NULL;
    }

    $sql = "UPDATE users SET username = :username WHERE id = :id LIMIT 1";
    $this->controller->db->getPdo()->beginTransaction();
    $this->controller->db->prepare($sql, [
      "username" => $name,
      "id"       => $this->controller->auth->getUserId()
    ]);

    $_SESSION[UserManager::SESSION_FIELD_USERNAME] = $name;

    return $this->controller->db->getPdo()->commit();
  }
```
I missed a session username refresh. Your session key tip is my solution.

**So this works well!**

---

Info: PHPStorm marks the method as internal, because the flag `@internal` is set in the class description.
![image](https://user-images.githubusercontent.com/15839724/58587862-a9ded300-825e-11e9-93b0-e1eb25de5b41.png)

---

I think a method like this can implement in the Auth framework. What do you think about this?

**Comment by ocram** [comment]
Thanks for trying this and for your feedback!

We will definitely add this feature to the library soon.

**Comment by PhantomArt** [comment]
I also find the username change function useful.

**Comment by kennyrkun** [comment]
Finding the lack of this feature quite disturbing...

**Comment by ocram** [comment]
It’s super late, but this little method is now part of the library:

 * https://github.com/delight-im/PHP-Auth/commit/7a20e966001ac6a3a8e6062a6411c36256b8f860
 * https://github.com/delight-im/PHP-Auth/commit/8875697eecda09e6cbbdb5d6a51eea0d8833e75a
 * https://github.com/delight-im/PHP-Auth/commit/5cc4745fc7aa85ef6a3fb0d40697bc45cb2cd3e9

**Comment by kennyrkun** [comment]
The incredible ocram delivers again! Thank you :3
